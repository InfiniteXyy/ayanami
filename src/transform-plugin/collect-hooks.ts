/* eslint-disable @typescript-eslint/no-non-null-assertion, @typescript-eslint/no-use-before-define, @typescript-eslint/no-unused-vars */
import * as ts from 'typescript'

export interface CollectHooksConfig {
  filter?: (filename: string) => boolean
  typeChecker?: ts.TypeChecker
}

export const collectAyanamiHooksFactory: (
  config?: CollectHooksConfig,
) => ts.TransformerFactory<ts.SourceFile> = (config = { filter: (_name: string) => true }) => (
  context,
) => {
  let typeChecker: ts.TypeChecker
  let sourceFile: ts.SourceFile
  const updateDeclaration = (
    parent: ts.FunctionDeclaration | ts.VariableStatement,
  ):
    | (ts.FunctionDeclaration | ts.VariableStatement)
    | [ts.FunctionDeclaration | ts.VariableStatement, ts.ExpressionStatement] => {
    let result: ts.FunctionDeclaration | ts.VariableDeclaration
    let finalResult: ts.FunctionDeclaration | ts.VariableStatement = parent
    let isMemo = false
    if (ts.isVariableStatement(parent)) {
      const declaration = parent.declarationList.declarations[0]
      const initializer = declaration.initializer!
      if (ts.isCallExpression(initializer)) {
        isMemo = true
      }
      result = declaration
    } else {
      result = parent
    }

    // Anonymous component
    if (!result.name) {
      return parent
    }

    const parameters = isMemo
      ? (((result as ts.VariableDeclaration).initializer! as ts.CallExpression)
          .arguments[0] as ts.FunctionExpression).parameters
      : ts.isFunctionDeclaration(result)
      ? result.parameters
      : (result.initializer! as ts.FunctionExpression).parameters

    // memo(forwardRef(memo(xxxxx))) WTF
    if (!parameters) {
      return parent
    }

    function addExpressionToComponent(expression: ts.ExpressionStatement) {
      if (ts.isFunctionDeclaration(result)) {
        result = ts.updateFunctionDeclaration(
          result,
          result.decorators,
          result.modifiers,
          result.asteriskToken,
          result.name,
          result.typeParameters,
          result.parameters,
          result.type,
          ts.updateBlock(result.body!, [expression, ...result.body!.statements]),
        )
        finalResult = result
      } else {
        if (ts.isArrowFunction(result.initializer!)) {
          if (ts.isBlock(result.initializer!.body)) {
            result = ts.updateVariableDeclaration(
              result,
              result.name,
              result.type,
              ts.updateArrowFunction(
                result.initializer!,
                result.initializer!.modifiers,
                result.initializer!.typeParameters,
                result.initializer!.parameters,
                result.initializer!.type,
                result.initializer!.equalsGreaterThanToken,
                ts.updateBlock(result.initializer!.body, [
                  expression,
                  ...result.initializer!.body.statements,
                ]),
              ),
            )
          } else {
            result = ts.updateVariableDeclaration(
              result,
              result.name,
              result.type,
              ts.updateArrowFunction(
                result.initializer!,
                result.initializer!.modifiers,
                result.initializer!.typeParameters,
                result.initializer!.parameters,
                result.initializer!.type,
                result.initializer!.equalsGreaterThanToken,
                ts.createBlock([expression, ts.createReturn(result.initializer!.body)], true),
              ),
            )
          }
        } else if (ts.isFunctionExpression(result.initializer!)) {
          result = ts.updateVariableDeclaration(
            result,
            result.name,
            result.type,
            ts.updateFunctionExpression(
              result.initializer!,
              result.initializer!.modifiers,
              result.initializer!.asteriskToken,
              result.initializer!.name,
              result.initializer!.typeParameters,
              result.initializer!.parameters,
              result.initializer!.type,
              ts.updateBlock(result.initializer!.body, [
                expression,
                ...result.initializer!.body.statements,
              ]),
            ),
          )
        } else if (ts.isCallExpression(result.initializer!)) {
          const paramFunc = result.initializer.arguments[0] as ts.FunctionExpression
          result = ts.updateVariableDeclaration(
            result,
            result.name,
            result.type,
            ts.updateCall(result.initializer, result.initializer.expression, void 0, [
              ts.updateFunctionExpression(
                paramFunc,
                paramFunc.modifiers,
                paramFunc.asteriskToken,
                paramFunc.name,
                paramFunc.typeParameters,
                paramFunc.parameters,
                paramFunc.type,
                ts.updateBlock(paramFunc.body, [expression, ...paramFunc.body.statements]),
              ),
            ]),
          )
        }
        const _parent = parent as ts.VariableStatement
        finalResult = ts.updateVariableStatement(
          _parent,
          parent.modifiers,
          ts.updateVariableDeclarationList(_parent.declarationList, [result]),
        )
      }
    }

    function addAdditionalExpression(identifier: ts.Expression) {
      if (!additionalResult) {
        additionalResult = ts.createExpressionStatement(
          ts.createBinary(
            ts.createElementAccess(
              ts.createIdentifier((result.name as ts.Identifier).text),
              ts.createPropertyAccess(
                ts.createCall(ts.createIdentifier('require'), undefined, [
                  ts.createStringLiteral('ayanami'),
                ]),
                ts.createIdentifier('SSRSymbol'),
              ),
            ),
            ts.SyntaxKind.EqualsToken,
            ts.createArrayLiteral([identifier], false),
          ),
        )
      } else {
        additionalResult = ts.updateExpressionStatement(
          additionalResult,
          ts.updateBinary(
            additionalResult.expression as ts.BinaryExpression,
            (additionalResult.expression as ts.BinaryExpression).left,
            ts.updateArrayLiteral(
              (additionalResult.expression as ts.BinaryExpression)
                .right as ts.ArrayLiteralExpression,
              [
                ...((additionalResult.expression as ts.BinaryExpression)
                  .right as ts.ArrayLiteralExpression).elements,
                identifier,
              ],
            ),
          ),
        )
      }
    }

    let additionalResult: ts.ExpressionStatement | undefined
    const functionLevelVisitor: ts.Visitor = (node) => {
      function warn() {
        console.warn(
          `Does not support run ayanami module with dynamic scope in SSR mode, ${
            ((node as ts.CallExpression).arguments[0] as ts.Identifier).text
          } will not run`,
        )
      }

      if (
        ts.isCallExpression(node) &&
        ts.isPropertyAccessExpression(node.expression) &&
        ts.isIdentifier(node.expression.expression) &&
        node.expression.expression.text === 'React' &&
        node.expression.name.text === 'createElement' &&
        Array.isArray(node.arguments) &&
        typeof node.arguments[0] !== 'undefined' &&
        ts.isIdentifier(node.arguments[0])
      ) {
        const componentIdentifier: ts.Identifier = node.arguments[0]
        const identifier = ts.createIdentifier(componentIdentifier.text)
        if (
          typeChecker
            .getSymbolsInScope(
              isMemo
                ? ((result as ts.VariableDeclaration).initializer! as ts.CallExpression)
                    .arguments[0]
                : parent,
              ts.SymbolFlags.Variable,
            )
            .every((symbol) => symbol.escapedName !== identifier.text) &&
          parameters.every((p) => p.name.getText(sourceFile) !== identifier.text)
        ) {
          addAdditionalExpression(
            ts.createElementAccess(
              identifier,
              ts.createPropertyAccess(
                ts.createCall(ts.createIdentifier('require'), undefined, [
                  ts.createStringLiteral('ayanami'),
                ]),
                ts.createIdentifier('SSRSymbol'),
              ),
            ),
          )
        }
      }

      if (
        ts.isCallExpression(node) &&
        ts.isIdentifier(node.expression) &&
        node.expression.text === 'useAyanami' &&
        Array.isArray(node.arguments) &&
        ts.isIdentifier(node.arguments[0])
      ) {
        if (node.arguments.length === 2) {
          const arg2 = node.arguments[1]
          if (ts.isObjectLiteralExpression(arg2)) {
            const scopeProperty = arg2.properties.find(
              (property) =>
                ts.isPropertyAssignment(property) &&
                ts.isIdentifier(property.name) &&
                property.name.text === 'scope',
            )
            if (scopeProperty) {
              const scopeAssignment = scopeProperty as ts.PropertyAssignment
              const initializer = scopeAssignment.initializer
              const moduleIdentifier = ts.createIdentifier(node.arguments[0].text)
              if (ts.isStringLiteral(initializer)) {
                addAdditionalExpression(
                  ts.createObjectLiteral([
                    ts.createPropertyAssignment(ts.createIdentifier('module'), moduleIdentifier),
                    ts.createPropertyAssignment(
                      ts.createIdentifier('scope'),
                      ts.createStringLiteral(initializer.text),
                    ),
                  ]),
                )
              } else {
                warn()
              }
            }
          } else {
            warn()
          }
        } else {
          const moduleIdentifier = ts.createIdentifier(node.arguments[0].text)
          addAdditionalExpression(moduleIdentifier)
        }
      }
      return ts.visitEachChild(node, functionLevelVisitor, context)
    }

    ts.visitEachChild(parent, functionLevelVisitor, context)

    if (additionalResult) {
      addExpressionToComponent(
        ts.createExpressionStatement(
          ts.createCall(
            ts.createPropertyAccess(
              ts.createCall(ts.createIdentifier('require'), undefined, [
                ts.createStringLiteral('ayanami'),
              ]),
              ts.createIdentifier('collectModules'),
            ),
            undefined,
            [
              ts.createElementAccess(
                ts.createIdentifier((result.name as ts.Identifier).text),
                ts.createPropertyAccess(
                  ts.createCall(ts.createIdentifier('require'), undefined, [
                    ts.createStringLiteral('ayanami'),
                  ]),
                  ts.createIdentifier('SSRSymbol'),
                ),
              ),
            ],
          ),
        ),
      )
      return [finalResult, additionalResult]
    }
    return finalResult
  }

  const topLevelvisitor: ts.Visitor = (node) => {
    if (node.parent && !ts.isSourceFile(node.parent)) {
      return node
    }

    if (
      (ts.isFunctionDeclaration(node) && node.body) ||
      (ts.isVariableStatement(node) &&
        node.declarationList &&
        Array.isArray(node.declarationList.declarations) &&
        ts.isVariableDeclaration(node.declarationList.declarations[0]) &&
        node.declarationList.declarations[0] &&
        node.declarationList.declarations[0].initializer &&
        (ts.isFunctionExpression(node.declarationList.declarations[0].initializer) ||
          ts.isArrowFunction(node.declarationList.declarations[0].initializer) ||
          isMemoComponent(node.declarationList.declarations[0].initializer)))
    ) {
      return updateDeclaration(node)
    }

    return ts.visitEachChild(node, topLevelvisitor, context)
  }

  return (node) => {
    if (node.languageVariant !== ts.LanguageVariant.JSX) {
      return node
    }

    if (!node.fileName) {
      return node
    }

    if (!config.filter!(node.fileName)) {
      return node
    }

    sourceFile = node

    if (!config.typeChecker) {
      const program = ts.createProgram({
        rootNames: [node.fileName],
        options: context.getCompilerOptions(),
      })
      typeChecker = program.getTypeChecker()
    } else {
      typeChecker = config.typeChecker
    }
    return ts.visitEachChild(node, topLevelvisitor, context)
  }
}

function isMemoComponent(node: ts.Node) {
  return (
    ts.isCallExpression(node) &&
    ((ts.isIdentifier(node.expression) && node.expression.text === 'memo') ||
      (ts.isPropertyAccessExpression(node.expression) &&
        ts.isIdentifier(node.expression.name) &&
        node.expression.name.text === 'memo' &&
        ts.isIdentifier(node.expression.expression) &&
        node.expression.expression.text === 'React'))
  )
}
