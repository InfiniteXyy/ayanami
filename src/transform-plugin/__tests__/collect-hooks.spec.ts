import * as ts from 'typescript'
import { join } from 'path'
import { readFileSync } from 'fs'

import { CollectHooksConfig, collectAyanamiHooksFactory } from '../collect-hooks'

function transpile(
  source: string,
  configs: {
    transpileOptions?: ts.TranspileOptions
    compilerOptions?: ts.CompilerOptions
    pluginOptions?: CollectHooksConfig
  } = {
    transpileOptions: {},
    compilerOptions: {},
    pluginOptions: undefined,
  },
) {
  return ts.transpileModule(source, {
    compilerOptions: {
      target: ts.ScriptTarget.ES5,
      jsx: ts.JsxEmit.React,
      moduleResolution: ts.ModuleResolutionKind.NodeJs,
      ...(configs.compilerOptions || {}),
    },
    transformers: {
      after: [collectAyanamiHooksFactory(configs.pluginOptions)],
    },
    ...(configs.transpileOptions || {}),
  }).outputText
}

function readFile(name: string) {
  return readFileSync(join(__dirname, `${name}.fixture`), 'utf-8')
}

describe('collect hooks plugin specs', () => {
  ;[
    {
      name: 'basic',
      describe: 'should do nothing to basic component',
    },
    {
      name: 'function',
    },
    {
      name: 'variable',
      options: {
        compilerOptions: {
          target: ts.ScriptTarget.ES2015,
        },
      },
    },
    {
      name: 'arrow-function-none-block',
      options: {
        compilerOptions: {
          target: ts.ScriptTarget.ES2015,
        },
      },
    },
    {
      name: 'variable-function',
      options: {
        compilerOptions: {
          target: ts.ScriptTarget.ES2015,
        },
      },
    },
    {
      name: 'literal-scope',
    },
    {
      name: 'dynamic-scope',
      describe: 'should ignore dynamic scope from props',
    },
    {
      name: 'component-from-props',
      describe: 'should ignore component from props',
    },
    {
      name: 'function-with-nonprops-argument',
      describe: 'should ignore component from params',
    },
    {
      name: 'anonymous-component',
      describe: 'should ignore anonymous component from props',
      options: {
        compilerOptions: {
          module: ts.ModuleKind.ESNext,
        },
      },
    },
    {
      name: 'smoke-1',
    },
    {
      name: 'smoke-2',
      options: {
        compilerOptions: {
          module: ts.ModuleKind.ESNext,
        },
      },
      describe: 'should not collect scoped component',
    },
    {
      name: 'with-memo',
      options: {
        compilerOptions: {
          module: ts.ModuleKind.ESNext,
        },
      },
    },
    {
      name: 'with-react-memo',
      options: {
        compilerOptions: {
          module: ts.ModuleKind.ESNext,
        },
      },
    },
    {
      name: 'with-memo-dynamic-scope',
      options: {
        compilerOptions: {
          module: ts.ModuleKind.ESNext,
        },
      },
      describe: 'should ignore memo with dynamic scope from props',
    },
  ].forEach((meta) => {
    it(meta.describe ? meta.describe : `should collect from ${meta.name} component`, () => {
      expect(transpile(readFile(meta.name), meta.options)).toMatchSnapshot()
    })
  })

  it('should do nothing if not jsx', () => {
    expect(transpile('function', { compilerOptions: { jsx: ts.JsxEmit.None } })).toBe('')
  })

  it('should do nothing if filename is been filter', () => {
    expect(
      transpile('function', {
        pluginOptions: { filter: () => false },
      }),
    ).toBe('')
  })
})
