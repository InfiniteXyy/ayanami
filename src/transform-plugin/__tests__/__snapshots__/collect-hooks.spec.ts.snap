// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collect hooks plugin specs should collect from arrow-function-none-block component 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { Modal } from 'antd';
const CountComponent = () => {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    return React.createElement(Modal, null);
};
CountComponent[require(\\"ayanami\\").SSRSymbol] = [Modal[require(\\"ayanami\\").SSRSymbol]];
ReactDOM.render(React.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should collect from function component 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var react_1 = require(\\"react\\");
var react_dom_1 = require(\\"react-dom\\");
var index_1 = require(\\"../../index\\");
function CountComponent() {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    var _a = index_1.useAyanami(Count), count = _a[0].count, actions = _a[1];
    var tips = index_1.useAyanami(Tips)[0].tips;
    var add = function (count) { return function () { return actions.add(count); }; };
    var minus = function (count) { return function () { return actions.minus(count); }; };
    return (react_1.default.createElement(\\"div\\", null,
        react_1.default.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        react_1.default.createElement(\\"p\\", null,
            \\"tips: \\",
            tips),
        react_1.default.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        react_1.default.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        react_1.default.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\")));
}
CountComponent[require(\\"ayanami\\").SSRSymbol] = [Count, Tips];
react_dom_1.default.render(react_1.default.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should collect from literal-scope component 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var react_1 = require(\\"react\\");
var react_dom_1 = require(\\"react-dom\\");
var index_1 = require(\\"../../index\\");
function CountComponent() {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    var _a = index_1.useAyanami(Count, { scope: 'Count' }), count = _a[0].count, actions = _a[1];
    var add = function (count) { return function () { return actions.add(count); }; };
    var minus = function (count) { return function () { return actions.minus(count); }; };
    return (react_1.default.createElement(\\"div\\", null,
        react_1.default.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        react_1.default.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        react_1.default.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        react_1.default.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\")));
}
CountComponent[require(\\"ayanami\\").SSRSymbol] = [{ module: Count, scope: \\"Count\\" }];
react_dom_1.default.render(react_1.default.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should collect from smoke-1 component 1`] = `
"var render = function (Component) {
    require(\\"ayanami\\").collectModules(render[require(\\"ayanami\\").SSRSymbol]);
    var _a = configureStore(), store = _a.store, history = _a.history;
    var client = InjectableFactory.getInstanceByToken(apolloClientProvider.provide);
    renderFunc(React.createElement(PrerenderCacheProvider, null,
        React.createElement(LocaleProvider, { locale: zhCN },
            React.createElement(Provider, { store: store },
                React.createElement(ApolloProvider, { client: client },
                    React.createElement(ConnectedRouter, { history: history },
                        React.createElement(Component, null)))))), document.querySelector('#app'));
};
render[require(\\"ayanami\\").SSRSymbol] = [PrerenderCacheProvider[require(\\"ayanami\\").SSRSymbol], LocaleProvider[require(\\"ayanami\\").SSRSymbol], Provider[require(\\"ayanami\\").SSRSymbol], ApolloProvider[require(\\"ayanami\\").SSRSymbol], ConnectedRouter[require(\\"ayanami\\").SSRSymbol]];
"
`;

exports[`collect hooks plugin specs should collect from variable component 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { Modal } from 'antd';
import { useAyanami } from '../../index';
const CountComponent = () => {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    const [{ count }, actions] = useAyanami(Count);
    const [{ tips }] = useAyanami(Tips);
    const add = (count) => () => actions.add(count);
    const minus = (count) => () => actions.minus(count);
    return (React.createElement(\\"div\\", null,
        React.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        React.createElement(\\"p\\", null,
            \\"tips: \\",
            tips),
        React.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        React.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        React.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\"),
        React.createElement(Modal, null)));
};
CountComponent[require(\\"ayanami\\").SSRSymbol] = [Count, Tips, Modal[require(\\"ayanami\\").SSRSymbol]];
ReactDOM.render(React.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should collect from variable-function component 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { useAyanami } from '../../index';
const CountComponent = function () {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    const [{ count }, actions] = useAyanami(Count);
    const [{ tips }] = useAyanami(Tips);
    const add = (count) => () => actions.add(count);
    const minus = (count) => () => actions.minus(count);
    return (React.createElement(\\"div\\", null,
        React.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        React.createElement(\\"p\\", null,
            \\"tips: \\",
            tips),
        React.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        React.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        React.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\")));
};
CountComponent[require(\\"ayanami\\").SSRSymbol] = [Count, Tips];
ReactDOM.render(React.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should collect from with-memo component 1`] = `
"export var Submissions = memo(function (props) {
    require(\\"ayanami\\").collectModules(Submissions[require(\\"ayanami\\").SSRSymbol]);
    var _a = useAyanami(SubmissionModule), state = _a[0], actions = _a[1];
    var loadMore = useCallback(function () {
        actions.nextPage();
    }, []);
    useEffect(function () {
        actions.getRecentSubmissions(props.userSlug);
        return actions.disposeFetching$;
    }, [props.userSlug]);
    return (React.createElement(FeedList, { activities: state.shownSubmissions, hasMore: state.currentPage < state.totalPage, isMyProfile: props.isMyProfile, loading: state.loading, loadMore: loadMore }));
});
Submissions[require(\\"ayanami\\").SSRSymbol] = [SubmissionModule, FeedList[require(\\"ayanami\\").SSRSymbol]];
"
`;

exports[`collect hooks plugin specs should collect from with-react-memo component 1`] = `
"export var Submissions = React.memo(function (props) {
    require(\\"ayanami\\").collectModules(Submissions[require(\\"ayanami\\").SSRSymbol]);
    var _a = useAyanami(SubmissionModule), state = _a[0], actions = _a[1];
    var loadMore = useCallback(function () {
        actions.nextPage();
    }, []);
    useEffect(function () {
        actions.getRecentSubmissions(props.userSlug);
        return actions.disposeFetching$;
    }, [props.userSlug]);
    return (React.createElement(FeedList, { activities: state.shownSubmissions, hasMore: state.currentPage < state.totalPage, isMyProfile: props.isMyProfile, loading: state.loading, loadMore: loadMore }));
});
Submissions[require(\\"ayanami\\").SSRSymbol] = [SubmissionModule, FeedList[require(\\"ayanami\\").SSRSymbol]];
"
`;

exports[`collect hooks plugin specs should do nothing to basic component 1`] = `
"function Button() {
    return React.createElement(\\"button\\", null, \\"click me\\");
}
"
`;

exports[`collect hooks plugin specs should ignore anonymous component from props 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { useAyanami } from '../../index';
export default function () {
    var _a = useAyanami(Count), count = _a[0].count, actions = _a[1];
    var tips = useAyanami(Tips)[0].tips;
    var add = function (count) { return function () { return actions.add(count); }; };
    var minus = function (count) { return function () { return actions.minus(count); }; };
    return (React.createElement(\\"div\\", null,
        React.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        React.createElement(\\"p\\", null,
            \\"tips: \\",
            tips),
        React.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        React.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        React.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\")));
}
ReactDOM.render(React.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should ignore component from params 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function TabTitle(props) {
    require(\\"ayanami\\").collectModules(TabTitle[require(\\"ayanami\\").SSRSymbol]);
    var icon = props.icon, Title = props.Title, toolTipTitle = props.toolTipTitle, isActive = props.isActive;
    var mainNode = (React.createElement(TabHeader, null,
        icon && React.createElement(Icon, { type: icon }),
        React.createElement(Title, { isActive: isActive })));
    return toolTipTitle ? React.createElement(Tooltip, { title: toolTipTitle }, mainNode) : mainNode;
}
TabTitle[require(\\"ayanami\\").SSRSymbol] = [TabHeader[require(\\"ayanami\\").SSRSymbol], Icon[require(\\"ayanami\\").SSRSymbol], Tooltip[require(\\"ayanami\\").SSRSymbol]];
exports.TabTitle = TabTitle;
"
`;

exports[`collect hooks plugin specs should ignore component from props 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var react_1 = require(\\"react\\");
var react_dom_1 = require(\\"react-dom\\");
var index_1 = require(\\"../../index\\");
function CountComponent(_a) {
    require(\\"ayanami\\").collectModules(CountComponent[require(\\"ayanami\\").SSRSymbol]);
    var Button = _a.Button;
    var _b = index_1.useAyanami(Count), count = _b[0].count, actions = _b[1];
    var tips = index_1.useAyanami(Tips)[0].tips;
    var add = function (count) { return function () { return actions.add(count); }; };
    var minus = function (count) { return function () { return actions.minus(count); }; };
    return (react_1.default.createElement(\\"div\\", null,
        react_1.default.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        react_1.default.createElement(\\"p\\", null,
            \\"tips: \\",
            tips),
        react_1.default.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        react_1.default.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        react_1.default.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\"),
        react_1.default.createElement(Button, { onClick: actions.send })));
}
CountComponent[require(\\"ayanami\\").SSRSymbol] = [Count, Tips];
react_dom_1.default.render(react_1.default.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should ignore dynamic scope from props 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
var react_1 = require(\\"react\\");
var react_dom_1 = require(\\"react-dom\\");
var index_1 = require(\\"../../index\\");
function CountComponent(_a) {
    var scope = _a.scope;
    var _b = index_1.useAyanami(Count, { scope: scope }), count = _b[0].count, actions = _b[1];
    var add = function (count) { return function () { return actions.add(count); }; };
    var minus = function (count) { return function () { return actions.minus(count); }; };
    return (react_1.default.createElement(\\"div\\", null,
        react_1.default.createElement(\\"p\\", null,
            \\"count: \\",
            count),
        react_1.default.createElement(\\"button\\", { onClick: add(1) }, \\"add one\\"),
        react_1.default.createElement(\\"button\\", { onClick: minus(1) }, \\"minus one\\"),
        react_1.default.createElement(\\"button\\", { onClick: actions.reset }, \\"reset to zero\\")));
}
react_dom_1.default.render(react_1.default.createElement(CountComponent, null), document.querySelector('#app'));
"
`;

exports[`collect hooks plugin specs should ignore memo with dynamic scope from props 1`] = `
"export var Submissions = memo(function (_a) {
    require(\\"ayanami\\").collectModules(Submissions[require(\\"ayanami\\").SSRSymbol]);
    var scope = _a.scope;
    var _b = useAyanami(SubmissionModule, {
        scope: scope,
    }), state = _b[0], actions = _b[1];
    var loadMore = useCallback(function () {
        actions.nextPage();
    }, []);
    useEffect(function () {
        actions.getRecentSubmissions(props.userSlug);
        return actions.disposeFetching$;
    }, [props.userSlug]);
    return (React.createElement(FeedList, { activities: state.shownSubmissions, hasMore: state.currentPage < state.totalPage, isMyProfile: props.isMyProfile, loading: state.loading, loadMore: loadMore }));
});
Submissions[require(\\"ayanami\\").SSRSymbol] = [FeedList[require(\\"ayanami\\").SSRSymbol]];
"
`;

exports[`collect hooks plugin specs should not collect scoped component 1`] = `
"export var SolutionTab = memo(function () {
    require(\\"ayanami\\").collectModules(SolutionTab[require(\\"ayanami\\").SSRSymbol]);
    var _a = useAyanami(QuestionDetailModule), state = _a[0], actions = _a[1];
    useEffect(function () {
        state.questionSlug && actions.getSolutionCount();
    }, [state.questionSlug]);
    var title = '题解' + (state.solutionCount ? \\"(\\" + state.solutionCount + \\")\\" : '');
    var Title = simpleTitle(title, '题解');
    return (React.createElement(React.Fragment, null,
        React.createElement(Icon, { type: \\"bottle\\" }),
        React.createElement(Title, null)));
});
SolutionTab[require(\\"ayanami\\").SSRSymbol] = [QuestionDetailModule, Icon[require(\\"ayanami\\").SSRSymbol]];
"
`;
